### DISCUSSION TITLE
User permissions for single DAG access (with task level details)

### DISCUSSION THREAD

**Discussion by fbo-cadran** [apache/airflow#30013] [Q&A] [answered]
I have an instance of Airflow 1.5.1 (webserver and scheduler) running without any problem. Multiple DAGs are specified and scheduled, and an Admin user is made for website interaction. However, I want a separate user with only access to specific DAGs. This user should be able to:

- Trigger this DAG via the website, and view DAG details (past run, etc).
- View task level details, see the DAG graph, view task logs.
- Be able to clear individual tasks. 

I wanted to use the most simplistic approach here, and go for the Access Control setup via the Webserver UI. As the admin user, I created a new role, and assigned **only**  this role to the new user. Based on the permission [documentation ](http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/latest/administration-and-deployment/security/access-control.html#permissions) and my listed requirements, I figured the following permissions are needed:

- DAG:<name>.can_read
- DAG:<name>.can_edit
- Website.can_read
- DAG Runs.can_read
- Task Instances.can_read
- Task Logs.can_read
- Task Instances.can_create
- DAG Runs.can_create

_Where name is replaced by the specific DAG in question._ 

This setup seems to work, mostly, and when logged in to the website, this user can only see this DAG. All actions that I want can be performed. However, I noted that when going to the task level details, so go example the /dags/<name>/grid, this user is able to access other DAGs when entering the name of the other dag. So, when specific urls are entered to view other DAG grids, or graphs, this user can view information about DAGs that should not be visible. 

This seems to be due to the task instances can read permission. I wonder if there is anything I forgot in this setup, or can change to achieve my goal.

Thanks in advance for the help!

**Comment by potiuk** [comment, marked as answer]
No. This is how it really works now I am afraid, the current implementation has only very rudimentary "dag level" permission support which is very limited and operates only on "DAG level" (as you noticed).

http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/latest/administration-and-deployment/security/access-control.html#dag-level-permissions

We are discussing more complete proposal (which will be formalized in AIP - Airflow Improvement Proposal). A discussion has been started recently on that on devlist:

* https://lists.apache.org/thread/srzsb8ppdnrrsj2osfm2gnqjh9fqzz6m 

And as result of it we changed the intial approach there and there is this github discussion that needs some comments and feedback https://github.com/apache/airflow/discussions/29986 - so if you feel like it, I think providing any comments and feedback there are welcome. This is all part of Airflow's multi-tenancy effortt that we build-in piece-by-piece. This piece is currently at the discussion phase.

**Reply by fbo-cadran** [reply]
Thank you for @potiuk for the explanation, looking forward to any changes in the future. Will keep a close eye on it.

**Reply by fbo-cadran** [reply]
One additional question @potiuk. Am I correct that DAGs.can_edit is set to a specific DAG, and the DAG Runs.can_create is given alongside, this user would only be able to create DAG runs on this specific DAG?

The reason I ask is because the working of the DAG-level permissions based on the documentation is currently still a bit unclear to me. As I understand you correctly, this is still limited and only operates on "DAG level". However, as listed on the documentation page, the get tasks website functionality needs the DAGs.can_read and Task Instances.can_read permission. Still, when the DAGs.can_read is specified to only one DAG, the tasks instances can read also has effect on other DAGs (as noted before). Therefore, I am unclear whether the DAG Runs.can_create permission also gives the opportunity to create other DAG Runs, for DAGs that are not listed on the DAGs.can_edit .

Thanks again for your time!

**Reply by potiuk** [reply]
I believe the idea is that you can't creat the DAGRun in this case. But don't believe my word here - i am not an oracle :). Try and see  and raise an issue if you find it is not what the docs say and how you interpret them.
